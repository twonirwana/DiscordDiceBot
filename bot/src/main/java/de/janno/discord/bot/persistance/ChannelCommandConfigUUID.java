package de.janno.discord.bot.persistance;

import lombok.NonNull;
import lombok.Value;

import java.time.LocalDateTime;
import java.util.UUID;

@Value
public class ChannelCommandConfigUUID {
    @NonNull
    String command;
    @NonNull
    LocalDateTime creationDate;
    @NonNull
    UUID configUUID;

}
