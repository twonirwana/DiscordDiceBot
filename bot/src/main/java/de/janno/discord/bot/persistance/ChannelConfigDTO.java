package de.janno.discord.bot.persistance;

import lombok.AllArgsConstructor;
import lombok.NonNull;
import lombok.Value;

import java.util.UUID;

@Value
@AllArgsConstructor
public class ChannelConfigDTO {

    @NonNull
    UUID configUUID;
    Long guildId;
    long channelId;
    Long userId;
    @NonNull
    String commandId;
    @NonNull
    String configClassId;
    @NonNull
    String config;
}
