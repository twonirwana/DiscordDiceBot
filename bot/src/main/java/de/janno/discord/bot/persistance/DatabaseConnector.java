package de.janno.discord.bot.persistance;

import com.google.common.collect.ImmutableSet;
import io.micrometer.core.instrument.binder.db.DatabaseTableMetrics;
import lombok.NonNull;
import org.h2.jdbcx.JdbcConnectionPool;
import org.jetbrains.annotations.Nullable;

import javax.sql.DataSource;
import java.sql.Connection;
import java.sql.SQLException;

import static io.micrometer.core.instrument.Metrics.globalRegistry;

public class DatabaseConnector {

    private final JdbcConnectionPool connectionPool;

    public DatabaseConnector(@NonNull String url, @Nullable String user, @Nullable String password) {
        connectionPool = JdbcConnectionPool.create(url, user, password);
        new DatabaseTableMetrics(connectionPool, "h2", "MESSAGE_DATA", ImmutableSet.of()).bindTo(globalRegistry);
        new DatabaseTableMetrics(connectionPool, "h2", "MESSAGE_CONFIG", ImmutableSet.of()).bindTo(globalRegistry);
        new DatabaseTableMetrics(connectionPool, "h2", "CHANNEL_CONFIG", ImmutableSet.of()).bindTo(globalRegistry);

    }

    public Connection getConnection() throws SQLException {
        return connectionPool.getConnection();
    }

    public void dispose() {
        connectionPool.dispose();
    }

    public DataSource getDataSource(){
        return connectionPool;
    }
}
