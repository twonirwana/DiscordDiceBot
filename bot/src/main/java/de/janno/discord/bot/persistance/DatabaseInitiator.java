package de.janno.discord.bot.persistance;

import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableList;
import com.google.common.io.Resources;
import io.micrometer.core.instrument.util.IOUtils;
import lombok.NonNull;
import lombok.Value;
import lombok.extern.slf4j.Slf4j;
import org.h2.jdbcx.JdbcConnectionPool;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.sql.*;
import java.time.LocalDateTime;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

@Slf4j
public class DatabaseInitiator {

    private final static List<String> MIGRATION_FILES = ImmutableList.<String>builder()
            .add("1_base.sql")
            .add("2_configTable.sql")
            .add("3_channelConfigUser.sql")
            .add("4_aliasCommandIdCleanUp.sql")
            .add("5_channelConfigFix.sql")
            .build();

    public static void initialize(@NonNull String url, @Nullable String user, @Nullable String password) {
        JdbcConnectionPool connectionPool = JdbcConnectionPool.create(url, user, password);

        List<Migration> allMigrations = readMigrations();
        log.info("found migrations: {}", allMigrations.stream().map(Migration::getName).collect(Collectors.joining(", ")));
        List<String> alreadyAppliedMigrations = getAlreadyAppliedMigrations(connectionPool);
        log.info("db migration status: {}", alreadyAppliedMigrations);
        applyMissingMigrations(allMigrations, alreadyAppliedMigrations, connectionPool);
    }

    private static void applyMissingMigrations(List<Migration> allMigrations, List<String> alreadyAppliedMigrations, JdbcConnectionPool connectionPool) {
        Preconditions.checkArgument(allMigrations.stream().map(Migration::getOrder).distinct().count() == allMigrations.size(), "Duplicate migration order");
        Preconditions.checkArgument(allMigrations.stream().map(Migration::getName).distinct().count() == allMigrations.size(), "Duplicate migration name");

        allMigrations.stream()
                .filter(m -> !alreadyAppliedMigrations.contains(m.getName()))
                .sorted(Comparator.comparing(Migration::getOrder))
                .forEach(m -> {
                    log.info("Start executing {}", m.getName());
                    try (Connection connection = connectionPool.getConnection()) {
                        Statement statement = connection.createStatement();
                        statement.execute(m.getSql());
                        try (PreparedStatement preparedStatement =
                                     connection.prepareStatement("INSERT INTO DB_VERSION(MIGRATION_NAME, CREATION_DATE) VALUES (?, ?)")) {
                            preparedStatement.setString(1, m.getName());
                            preparedStatement.setTimestamp(2, Timestamp.valueOf(LocalDateTime.now()));
                            preparedStatement.execute();
                        }
                    } catch (SQLException e) {
                        log.error("Error in {}", m.getName());
                        throw new RuntimeException(e);
                    }
                    log.info("Finish executing {}", m.getName());
                });
    }

    private static List<String> getAlreadyAppliedMigrations(JdbcConnectionPool connectionPool) {
        try (Connection connection = connectionPool.getConnection()) {
            Statement statement = connection.createStatement();
            ResultSet resultSet = statement.executeQuery("select MIGRATION_NAME from DB_VERSION");
            ImmutableList.Builder<String> builder = ImmutableList.builder();
            while (resultSet.next()) {
                builder.add(resultSet.getString("MIGRATION_NAME"));
            }
            return builder.build();
        } catch (SQLException e) {
            return List.of();
        }
    }

    private static List<Migration> readMigrations() {
        return MIGRATION_FILES.stream()
                .map(DatabaseInitiator::readMigration)
                .toList();
    }

    private static Migration readMigration(final String fileName) {
        Preconditions.checkArgument(fileName.contains("_"), "Wrong file format: {}", fileName);
        Preconditions.checkArgument(fileName.endsWith("sql"), "Wrong file format: {}", fileName);
        String content;
        try {
            content = IOUtils.toString(Resources.getResource("db/migrations/" + fileName).openStream(), StandardCharsets.UTF_8);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        int firstUnderscoreIndex = fileName.indexOf("_");
        String order = fileName.substring(0, firstUnderscoreIndex);
        String name = fileName.substring(firstUnderscoreIndex + 1, fileName.length() - 4);
        return new Migration(order, name, content);
    }


    @Value
    private static class Migration {
        String order;
        String name;
        String sql;
    }
}
