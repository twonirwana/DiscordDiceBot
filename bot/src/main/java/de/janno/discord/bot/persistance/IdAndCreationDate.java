package de.janno.discord.bot.persistance;

import lombok.Value;

import java.time.LocalDateTime;

@Value
public class IdAndCreationDate {
        long id;
        LocalDateTime creationDate;
}
