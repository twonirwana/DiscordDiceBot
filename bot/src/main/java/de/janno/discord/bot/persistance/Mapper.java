package de.janno.discord.bot.persistance;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
import lombok.NonNull;


public class Mapper {

    public static final String NO_PERSISTED_STATE = "None";
    private static final ObjectMapper mapper = new ObjectMapper(new YAMLFactory());

    public static String serializedObject(@NonNull Object object) {
        try {
            return mapper.writeValueAsString(object);
        } catch (JsonProcessingException e) {
            throw new IllegalStateException(e);
        }
    }

    public static <T> T deserializeObject(@NonNull String object, @NonNull Class<T> classOfObject) {
        try {
            return mapper.readValue(object, classOfObject);
        } catch (JsonProcessingException e) {
            throw new IllegalStateException(e);
        }
    }

}
