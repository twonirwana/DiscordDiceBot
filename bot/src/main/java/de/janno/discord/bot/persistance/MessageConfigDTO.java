package de.janno.discord.bot.persistance;

import lombok.AllArgsConstructor;
import lombok.NonNull;
import lombok.Value;
import javax.annotation.Nullable;

import java.util.UUID;

@Value
@AllArgsConstructor
public class MessageConfigDTO {

    @NonNull
    UUID configUUID;
    @Nullable
    Long guildId;
    long channelId;
    @NonNull
    String commandId;
    @NonNull
    String configClassId;
    @NonNull
    String config;

}
