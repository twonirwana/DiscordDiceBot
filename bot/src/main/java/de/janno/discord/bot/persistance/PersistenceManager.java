package de.janno.discord.bot.persistance;

import lombok.NonNull;

import java.util.Optional;
import java.util.Set;
import java.util.UUID;

public interface PersistenceManager {

    @NonNull Optional<MessageConfigDTO> getMessageConfig(@NonNull UUID configUUID);

    @NonNull Optional<MessageConfigDTO> getConfigFromMessage(long channelId, long messageId);

    void saveMessageConfig(@NonNull MessageConfigDTO messageConfigDTO);

    @NonNull Optional<MessageDataDTO> getMessageData(long channelId, long messageId);

    @NonNull Set<Long> getAllMessageIdsForConfig(@NonNull UUID configUUID);

    void deleteStateForMessage(long channelId, long messageId);

    @NonNull Set<Long> deleteMessageDataForChannel(long channelId);

    void saveMessageData(@NonNull MessageDataDTO messageState);

    Set<Long> getAllGuildIds();

    @NonNull Optional<ChannelConfigDTO> getChannelConfig(long channelId, String configClassId);

    void saveChannelConfig(@NonNull ChannelConfigDTO channelConfigDTO);

    void deleteChannelConfig(long channelId, String configClassId);
}
