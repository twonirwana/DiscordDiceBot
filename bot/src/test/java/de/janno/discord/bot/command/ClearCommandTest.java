package de.janno.discord.bot.command;

import com.google.common.collect.ImmutableSet;
import de.janno.discord.bot.persistance.PersistenceManager;
import de.janno.discord.connector.api.SlashEventAdaptor;
import org.junit.jupiter.api.Test;
import reactor.core.publisher.Mono;
import reactor.test.StepVerifier;

import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.*;

class ClearCommandTest {

    final PersistenceManager persistenceManager = mock(PersistenceManager.class);
    final ClearCommand underTest = new ClearCommand(persistenceManager);

    @Test
    void getName() {
        assertThat(underTest.getCommandId()).isEqualTo("clear");
    }

    @Test
    void handleSlashCommandEvent() {
        SlashEventAdaptor slashEventAdaptor = mock(SlashEventAdaptor.class);
        when(slashEventAdaptor.reply(any(), anyBoolean())).thenReturn(Mono.empty());
        when(slashEventAdaptor.getChannelId()).thenReturn(0L);
        when(persistenceManager.deleteMessageDataForChannel(anyLong())).thenReturn(ImmutableSet.of(1L, 2L));
        when(slashEventAdaptor.deleteMessageById(anyLong())).thenReturn(Mono.empty());

        Mono<Void> res = underTest.handleSlashCommandEvent(slashEventAdaptor, () -> UUID.fromString("00000000-0000-0000-0000-000000000000"));
        StepVerifier.create(res).verifyComplete();


        verify(persistenceManager).deleteMessageDataForChannel(0L);
        verify(slashEventAdaptor).deleteMessageById(1L);
        verify(slashEventAdaptor).deleteMessageById(2L);
    }

}