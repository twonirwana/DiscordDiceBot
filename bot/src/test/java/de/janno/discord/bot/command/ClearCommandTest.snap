de.janno.discord.bot.command.ClearCommandTest.getCommandDefinition=[
  {
    "name" : "clear",
    "description" : "Removes all button messages and saved bot data for this channel",
    "nameLocales" : [ ],
    "descriptionLocales" : [
      {
        "locale" : "de",
        "description" : "Entferne alle Button-Nachrichten und lösche alle dazugehörigen Daten in diesen Kanal"
      },
      {
        "locale" : "pt_BR",
        "description" : "Remove todas mensagens de botões e dados salvos pelo bot nesse canal"
      }
    ],
    "options" : [ ]
  }
]


de.janno.discord.bot.command.ClearCommandTest.getId=[
  "clear"
]